### PR TITLE
[#735] Detect and reconcile alarms updated while mapper was down

### PR DESCRIPTION
**Pull request by albinsuresh**
## Proposed changes

Detect and reconcile any alarms that the mapper may have missed while it was down, on its restart. The mapper needs to detect any newly raised  alarms as well as the ones that were cleared while it was down.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: crates/common/download/src/download.rs [hunk 1]
```diff
@@ -1,9 +1,8 @@
 use crate::error::DownloadError;
 use backoff::{future::retry, ExponentialBackoff};
-use nix::{
-    fcntl::{fallocate, FallocateFlags},
-    sys::statvfs,
-};
+#[cfg(target_os = "linux")]
```
Comment (albinsuresh) [inline crates/common/download/src/download.rs:3]: The changes in this file were made just to make sure the project compile on my non-Linux dev machine. I could move it to a dedicated PR if it's annoying. Left it in this PR itself as the changes were limited to just this file,

File: crates/core/tedge_mapper/src/mapper.rs [hunk 2]
```diff
@@ -85,14 +89,121 @@ impl Mapper {
             let _ = self.output.send(init_message).await;
         }
 
-        while let Some(message) = &mut self.input.next().await {
-            let converted_messages = self.converter.convert(&message);
-            for converted_message in converted_messages.into_iter() {
-                let _ = self.output.send(converted_message).await;
-            }
+        // Detect and reconcile any alarms that this mapper may have missed while not running
+        self.reconcile_missed_alarms().await;
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/mapper.rs:93]: Beware that all the mapper processes are running this code (c8y, az and collectd).

So I would delegate this to the converter. I agree that's not easy: the converter can have a state but should not be given a direct access to mqtt.

File: crates/core/tedge_mapper/src/mapper.rs [hunk 3]
```diff
@@ -85,14 +89,121 @@ impl Mapper {
...
+                    if message.topic.name.starts_with("tedge/alarms") {
+                        let alarm_id = message
+                            .topic
+                            .name
+                            .strip_prefix("tedge/alarms/")
+                            .expect("Expected tedge/alarms prefix")
+                            .to_string();
+                        pending_alarms_map
+                            .insert(alarm_id.clone(), message);
+                    } else if message.topic.name.starts_with("tedge-internal/alarms") {
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/mapper.rs:143]: The internal topic name should be specific to the cloud end-endpoint, in case the device is connected to two different end-points.

File: crates/core/tedge_mapper/src/mapper.rs [hunk 4]
```diff
@@ -85,14 +89,121 @@ impl Mapper {
...
+                    .expect("Expected tedge/alarms prefix")
+                    .to_string();
+
+                let topic =
+                    Topic::new_unchecked(format!("tedge-internal/alarms/{}", alarm_id).as_str());
+                let alarm_copy =
+                    Message::new(&topic, message.payload_bytes().to_owned()).with_retain();
+                let _ = self.output.send(alarm_copy).await;
+            }
+        }
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/mapper.rs:205]: This needs to be moved inside the converter.

When an alarm is received, the converter must return *two* messages:
* the translated alarm,
* and the internal copy.

File: crates/core/tedge_mapper/src/mapper.rs [hunk 5]
```diff
@@ -85,14 +89,121 @@ impl Mapper {
...
+            tokio::select! {
+                // When the reconciliation window times out, break out of the select loop to stop caching alarm messages
+                _ = &mut reconciliation_window => {
+                    break;
+                }
+
+                // During the reconciliation window, cache all the alarm messages received from tedge/alarms and tedge-internal alarms topics.
+                // This window gives the mapper ample time to receive all the retained messages from both the topics.
+                Some(message) = self.input.next() => {
+                    if message.topic.name.starts_with("tedge/alarms") {
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/mapper.rs:134]: Beware, that we are consuming *all* types of messaging here, but processing *only* those related to alarms.

One needs:
* either to process all non-alarm messages with `self.converter.convert(&message)`,
* or to use a different connection and session name for this reconciliation step.

I have a preference for the second alternative, because doing so:
* The reconciliation loop focuses on alarms.
* The main process loop doesn't need to subscribe to the `tedge-internal/alarms/` alarm topic.

File: crates/core/tedge_mapper/src/c8y_converter.rs [hunk 6]
```diff
@@ -114,6 +117,9 @@ impl Converter for CumulocityConverter {
             self.try_convert_measurement(input)
         } else if input.topic.name.starts_with("tedge/alarms") {
             self.try_convert_alarm(input)
+        } else if input.topic.name.starts_with("tedge-internal/alarms") {
+            // Do not try to convert as this is an internal message
+            Ok(vec![])
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y_converter.rs:122]: This case can be removed. If the reconciliation loop is rewritten to use its own MQTT connection, then the main loop can avoid to subscribe to the `tedge-internal/alarms` topic.

**Review by didier-wenzek — COMMENTED** [review]
This is a great move in the right direction. However:
* The init and main loops need to be better separate. With the current version of the code messages are lost during the reconciliation step.
* The converter should take responsibility for all message conversions, including alarms. Currently all the non-c8y mappers will incorrectly reconcile the alarms. 

Thera are many spelling variants for reconciliation ;-):
* Reconiciliation
* reconcicilation

**Comment by didier-wenzek** [comment]
The test `mapper::tests::a_valid_input_leads_to_a_translated_output` ... FAILED because of the added 3 seconds of reconciliation.

```
---- mapper::tests::a_valid_input_leads_to_a_translated_output stdout ----
MQTT-TEST INFO: start test MQTT broker (port = 55555)
MQTT-TEST MSG: topic = in_topic, payload = "abcde"
thread 'mapper::tests::a_valid_input_leads_to_a_translated_output' panicked at 'assertion failed: `(left == right)`
  left: `Some("ABCDE")`,
 right: `None`', crates/core/tedge_mapper/src/mapper.rs:251:9
```

File: crates/core/tedge_mapper/src/mapper.rs [hunk 7]
```diff
@@ -85,14 +89,121 @@ impl Mapper {
...
+                            .name
+                            .strip_prefix("tedge-internal/alarms/")
+                            .expect("Expected tedge-internal/alarms prefix")
+                            .to_string();
+                        old_alarms_map
+                            .insert(alarm_id, message);
+                    }
+                }
+            }
+        }
```
Comment (makr11st) [inline crates/core/tedge_mapper/src/mapper.rs:155]: I would rather use `tokio::timeout` to drive this instead of manually pushing the `select` loop, 

```suggestion
        let _ = tokio::time::timeout(Duration::from_secs(3), async {
            while let Some(message) = self.input.next().await {
                if message.topic.name.starts_with("tedge/alarms") {
                    let alarm_id = message
                        .topic
                        .name
                        .strip_prefix("tedge/alarms/")
                        .expect("Expected tedge/alarms prefix")
                        .to_string();
                    pending_alarms_map.insert(alarm_id.clone(), message);
                } else if message.topic.name.starts_with("tedge-internal/alarms") {
                    let alarm_id = message
                        .topic
                        .name
                        .strip_prefix("tedge-internal/alarms/")
                        .expect("Expected tedge-internal/alarms prefix")
                        .to_string();
                    old_alarms_map.insert(alarm_id, message);
                }
            }
        })
        .await;
```

File: crates/core/tedge_mapper/src/mapper.rs [hunk 8]
```diff
@@ -85,14 +89,121 @@ impl Mapper {
...
+
+                // If the payload of a message received from tedge/alarms is same as one received from tedge-internal/alarms,
+                // it is assumed to be one that was already processed earlier and hence removed from the pending alarms list.
+                Some(message) => {
+                    if message.payload_bytes() == old_message.payload_bytes() {
+                        pending_alarms_map.remove(&key);
+                    }
+                }
+            }
+        }
```
Comment (makr11st) [inline crates/core/tedge_mapper/src/mapper.rs:177]: An alternative implementation:
```suggestion
        old_alarms_map.iter().for_each(|(alarm_id, message)| {
            match pending_alarms_map.entry(alarm_id.to_string()) {
                Entry::Occupied(e) => {
                    if message.payload_bytes() == e.get().payload_bytes() {
                        e.remove();
                    }
                }

                Entry::Vacant(e) => {
                    let topic = Topic::new_unchecked(format!("tedge/alarms/{}", e.key()).as_str());
                    let message = Message::new(&topic, vec![]).with_retain();
                    e.insert(message);
                }
            }
        });
```

I hope I got the logic right.

File: crates/core/tedge_mapper/src/mapper.rs [hunk 9]
```diff
@@ -85,14 +89,121 @@ impl Mapper {
...
+
+        // Once all the pending alarms are identified, process them
+        for (_key, message) in pending_alarms_map.into_iter() {
+            self.process_message(message).await;
+        }
```
Comment (makr11st) [inline crates/core/tedge_mapper/src/mapper.rs:182]: I believe it would be better to just return the vec of Message and send it in the main loop altogether with other init messages maybe?

File: crates/core/tedge_mapper/src/mapper.rs [hunk 10]
```diff
@@ -85,14 +89,121 @@ impl Mapper {
...
+    /// All the live alarms are received from tedge/alarms topic on startup.
+    /// Similarly, all the previously processed alarms are received from tedge-internal/alarms topic.
+    /// Reconiciliation detects the difference between these two sets, which are the missed messages.
+    ///
+    /// An alarm that is present in tedge-internal/alarms, but not in tedge/alarms topic
+    /// is assumed to have been cleared while the mapper process was down.
+    /// Similarly, an alarm that is present in tedge/alarms, but not in tedge-internal/alarms topic
+    /// is one that was raised while the mapper process was down.
+    /// An alarm present in both, if their payload is the same, is one that was already processed before the restart
+    /// and hence can be ignored during reconcicilation.
```
Comment (makr11st) [inline crates/core/tedge_mapper/src/mapper.rs:116]:
```suggestion
    /// and hence can be ignored during reconciliation.
```

**Review by makr11st — COMMENTED** [review]
I would like to make a suggestion regarding the name, the word `reconcile` is much less often used and I believe what we are doing here is a simple `synchronisation`, could we maybe use the abbreviation `sync`, I'm proposing this for 2 reasons:

1. It is easier to read the code and documentation when we use simpler wording and I believe the word `sync` is more common and known then the word `reconcile`/`reconciliation` (I had to look up how to correctly spell it).
2. In the PR you have made a few spelling mistakes with this word, why make our lives harder?

File: crates/core/tedge_mapper/src/mapper.rs (hunk 7 above)
Comment (albinsuresh) [reply crates/core/tedge_mapper/src/mapper.rs:155]: This might work, but I don't know if the usage of `timeout` is semantically correct here. The documentation of `tokio::timeout` states that it will cancel the underlying future of the task doesn't finish within the given timeframe. So, it kinda indicates a failure. But as per our logic, what we need is more of a **window** during which a task is performed. So, for a reader, using a timeout here might be a bit confusing, I feel. But I agree that it makes the code much simpler. So, not fully sure. @didier-wenzek WDYT?

BTW, I don't know what the implication of that future cancellation is. Hope there is none as our task just populates some maps and doesn't do anything blocking.

Comment (makr11st) [reply crates/core/tedge_mapper/src/mapper.rs:155]: That's good point, I think that first the work proposed by Didier should be done to move this part to the converter this part my not even be required at all.

File: crates/core/tedge_mapper/src/mapper.rs (hunk 2 above)
Comment (albinsuresh) [reply crates/core/tedge_mapper/src/mapper.rs:93]: This would be the heart of the new design and here's how I'm planning to do this. I'll have to introduce 3 states that the converter can be in:
1. Caching phase: Where all alarms received on the startup , from the tedge/alarms and the internal/alarms topics, are cached into their respective maps without doing an immediate conversion. Messages received on other topics will be simply converted, as they would be during the normal phase as well.
2. Reconciliation phase: Both the maps produced during the caching phase are compared and the differences are converted and sent to the respective cloud topics.
3. The normal phase: This is the regular phase during which the converter would just be doing the simple stateless message mapping.

The mapper main would be the one driving the converter through these stages based on the timer (the reconciliation window). The converter would have no notion of time. It just sees itself in different states and reacts accordingly.

File: crates/core/tedge_mapper/src/mapper.rs (hunk 7 above)
Comment (didier-wenzek) [reply crates/core/tedge_mapper/src/mapper.rs:155]: I think that the two alternatives are equivalent regarding cancellation of the message loop. Hence, using a timeout will simplify the code.

File: crates/core/tedge_mapper/src/mapper.rs (hunk 5 above)
Comment (didier-wenzek) [reply crates/core/tedge_mapper/src/mapper.rs:134]: My idea was to use two connections in sequence. A first one that only reads alarms and internal alarms and stop when all the retained messages has been received. A second one that reads alarms and measurements. I agree that there is window where we can have a race.

I'm not convinced that a single connection is fool-proof. Sure, this remove the race condition on cleared alarms; but this also make things complicated/fragile. For instance, the current version of the code forget to process the measurements during the reconciliation loop.

Having two parallel connections seems better to me. No race condition on cleared alarms and no risk to have regressions when we update the alarm or measurement processing. We don't have to stop listening for internal topics , we can simple ignore them - as done is a solution with a single connection.

File: crates/core/tedge_mapper/src/mapper.rs (hunk 2 above)
Comment (didier-wenzek) [reply crates/core/tedge_mapper/src/mapper.rs:93]: Your proposal sounds good.

My remark was more about the fact that this `reconcile_missed_alarms` loop is called by *all* the mappers and that *only* the c8y mapper should reconcile alarms for c8y. With the current code base, the collectd mapper will also reconcile the alarms ... its own way, meaning that the internal alarms will be updated but nothing forwarded.

File: crates/core/tedge_mapper/src/mapper.rs (hunk 9 above)
Comment (albinsuresh) [reply crates/core/tedge_mapper/src/mapper.rs:182]: I don't think combining these with the init messages would be a good idea because the init messages need to go to the cloud as soon as the mapper starts, whereas these sync messages will be delayed by a few seconds(the synchronization window) during which the non-alarm messages still need to be converted and sent to the cloud. So, I guess they have to be separate.

File: crates/core/tedge_mapper/src/c8y_converter.rs [hunk 11]
```diff
@@ -18,12 +19,18 @@ const INVENTORY_FRAGMENTS_FILE_LOCATION: &str = "/etc/tedge/device/inventory.jso
...
 
 pub struct CumulocityConverter {
     pub(crate) size_threshold: SizeThreshold,
     children: HashSet<String>,
     pub(crate) mapper_config: MapperConfig,
     device_name: String,
+
+    syncing: bool,
+    pending_alarms_map: HashMap<String, Message>,
+    old_alarms_map: HashMap<String, Message>,
```
Comment (didier-wenzek) [inline crates/core/tedge_mapper/src/c8y_converter.rs:33]: This 3 fields are always used in concert with a systematic pattern `if syncing {do this with the maps} else {do that, the maps being used}`.

=> I would extract these fields in a separate struct and use an enum to differentiate the cases.

```
pub struct CumulocityConverter {
     ... 
     synchronizer: AlarmSynchronizer,
}

enum AlarmSynchronizer {
     Syncing { pending_alarms_map: HashMap<String, Message>, old_alarms_map: HashMap<String, Message> }
     SyncDone
}

impl AlarmSynchronizer  {
     fn try_convert_alarm(&mut self, input: &Message) -> Result<Vec<Message>, ConversionError> { }
     fn process_internal_alarm(&mut self, input: &Message) {}
     fn sync_messages(&mut self ) {}
}
```

File: crates/core/tedge_mapper/src/mapper.rs [hunk 12]
```diff
@@ -85,14 +87,34 @@ impl Mapper {
             let _ = self.output.send(init_message).await;
         }
 
-        while let Some(message) = &mut self.input.next().await {
-            let converted_messages = self.converter.convert(&message);
-            for converted_message in converted_messages.into_iter() {
-                let _ = self.output.send(converted_message).await;
+        let sync_window = Duration::from_secs(3);
```
Comment (makr11st) [inline crates/core/tedge_mapper/src/mapper.rs:90]: This could be a const.

File: crates/core/tedge_mapper/src/c8y_converter.rs [hunk 13]
```diff
@@ -129,6 +188,58 @@ impl Converter for CumulocityConverter {
...
+    /// Similarly, all the previously processed alarms are received from c8y-internal/alarms topic.
+    /// Reconiciliation detects the difference between these two sets, which are the missed messages.
+    ///
+    /// An alarm that is present in c8y-internal/alarms, but not in tedge/alarms topic
+    /// is assumed to have been cleared while the mapper process was down.
+    /// Similarly, an alarm that is present in tedge/alarms, but not in c8y-internal/alarms topic
+    /// is one that was raised while the mapper process was down.
+    /// An alarm present in both, if their payload is the same, is one that was already processed before the restart
+    /// and hence can be ignored during reconcicilation.
+    fn sync_messages(&mut self) -> Vec<Message> {
```
[the remark this answers is not in the document]
Comment (albinsuresh) [reply crates/core/tedge_mapper/src/c8y_converter.rs:206]: It couldn't be local variables in this function because it's the `convert` methods that populates those maps and this `sync` method just consumes these populated maps. So I needed these maps shared between these functions and that's why it was kept as fields of the `Converter`.

But you're right that we don't need to keep these maps beyond the sync method call and that's why I had used drain. In interpreted the "leaves the memory allocated for later reuse" as it just leaves the memory for an empty map allocated. Didn't expect them to leave the memory for **all entries** allocated for later reuse.

[the remark this answers is not in the document]
Comment (makr11st) [reply crates/core/tedge_mapper/src/c8y_converter.rs:206]: That's fine, as an alternative it would be possible that one of these methods takes a generator function which takes a map of messages and produces vec of messages to sent.
I could provide example code later on, though.

I'm curious to see the enum impl.

**Review by didier-wenzek — COMMENTED** [review]
This looks good. Using an enum to encode the synchronization state clarifies the code a lot. Unfortunately, the two new tests are failing.

**Review by didier-wenzek — COMMENTED** [review]
Approved
